README.md
# text2image

[![npm](https://img.shields.io/npm/v/@xg4/text2image.svg)](https://www.npmjs.com/package/@xg4/text2image)
[![npm](https://img.shields.io/npm/l/@xg4/text2image.svg)](https://www.npmjs.com/package/@xg4/text2image)

> convert text to image by canvas

## Installation

### Install with npm or Yarn

```sh
# npm
npm i @xg4/text2image

# yarn
yarn add @xg4/text2image

# pnpm
pnpm i @xg4/text2image
```

## Usage

### Instance

```js
import text2image, { Text2Image } from '@xg4/text2image'

// use current options convert default options
text2image.setDefaultOptions({
  // some options
})

// reset default options
text2image.resetDefaultOptions()

// you can create new instance or use text2image singleton
const t2 = new Text2Image()
// or
// initialization default options
const t2 = new Text2Image({
  fontSize: 13,
  color: '#000000',
  fontFamily: 'arial',
  fontWeight: 'bold',
  type: 'image/png',
  quality: 0.92,
})

// use current options convert default options
t2.setDefaultOptions({
  // some options
})

// reset default options
t2.resetDefaultOptions()
```

## Object URL

### js/html

```js
async function renderTextImage() {
  const url = await text2image.createURL('hello world')
  const img = new Image()
  img.src = url
  document.body.appendChild(img)
}

renderTextImage()
```

### React

```tsx
import { useEffect, useState } from 'react'
import text2image from '../../../src'

export default function App() {
  const [images, setImages] = useState<string[]>([])

  useEffect(() => {
    async function renderTextImage() {
      const url = await text2image.createURL('Hello World')
      setImages((urls) => [...urls, url])
    }

    renderTextImage()
  }, [])

  return (
    <div>
      {images.map((url, index) => (
        <img
          src={url}
          key={index}
          onLoad={() => {
            // img loaded, should to destroy object url
            text2image.destroyURL(url)
          }}
        />
      ))}
    </div>
  )
}
```

### mask image

```js
async function renderTextImage() {
  // get mask image
  await text2image.loadImage(imgUrl)

  // create object url
  const url = await text2image.createURL('hello world')
  // or
  // const url = await text2image.createURL({
  //   text: 'hello world',
  //   // some options
  // })

  const img = new Image()
  // img loaded, should to destroy object url
  img.onload = () => {
    text2image.destroyURL(url)
  }
  img.src = url

  document.body.appendChild(img)
}

renderTextImage()
```

## DataURL

### js/html

```js
const url = text2image.toDataURL('hello world')
const img = new Image()
img.src = url
document.body.appendChild(img)
```

### React

```tsx
import { useEffect, useState } from 'react'
import text2image from '../../../src'

export default function App() {
  const [images, setImages] = useState<string[]>([])

  useEffect(() => {
    const url = text2image.toDataURL('Hello World')
    // const url = text2image.toDataURL({
    //   text: 'hello world',
    //   // some options
    // })
    setImages((urls) => [...urls, url])
  }, [])

  return (
    <div>
      {images.map((url, index) => (
        <img src={url} key={index} />
      ))}
    </div>
  )
}
```

### mask image

```js
async function getTextImage() {
  // get mask image
  await text2image.loadImage(imgUrl)

  // create data url
  const url = text2image.toDataURL('hello world')
  // or
  // const url = text2image.toDataURL({
  //   text: 'hello world',
  //   // some options
  // })

  const img = new Image()
  img.src = url

  document.body.appendChild(img)
}

getTextImage()
```

## Options

| name         | type             | default     | description                  |
| ------------ | ---------------- | ----------- | ---------------------------- |
| `text`       | `string`         | `null`      | image content                |
| `fontSize`   | `number\|string` | `30`        | font size(like css)          |
| `fontWeight` | `number\|string` | `normal`    | font weight(like css)        |
| `fontFamily` | `string`         | `arial`     | font family(like css)        |
| `color`      | `string`         | `#000000`   | font color(like css)         |
| `type`       | `string`         | `image/png` | image type                   |
| `quality`    | `number`         | `0.92`      | image quality                |
| `alpha`      | `number`         | `0.3`       | mask alpha(水印图片的透明度) |

## Example

> [https://text-image.vercel.app/](https://text-image.vercel.app/)

## Contributing

Welcome

- Fork it

- Submit pull request

## Polyfills needed to support older browsers

> `HTMLCanvasElement.prototype.toBlob`: see [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toBlob#Polyfill) for details about unsupported older browsers and a simple polyfill.

```js
;(function () {
  if (!HTMLCanvasElement.prototype.toBlob) {
    Object.defineProperty(HTMLCanvasElement.prototype, 'toBlob', {
      value: function (callback, type, quality) {
        var binStr = atob(this.toDataURL(type, quality).split(',')[1]),
          len = binStr.length,
          arr = new Uint8Array(len)

        for (var i = 0; i < len; i++) {
          arr[i] = binStr.charCodeAt(i)
        }

        callback(new Blob([arr], { type: type || 'image/png' }))
      },
    })
  }
})()
```

## Browsers support

Modern browsers and IE10.

| [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png" alt="IE / Edge" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>IE / Edge | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png" alt="Firefox" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Firefox | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png" alt="Chrome" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Chrome | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png" alt="Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Safari | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/opera/opera_48x48.png" alt="Opera" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)</br>Opera |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| IE10, Edge                                                                                                                                                                                                      | last 2 versions                                                                                                                                                                                                   | last 2 versions                                                                                                                                                                                               | last 2 versions                                                                                                                                                                                               | last 2 versions                                                                                                                                                                                           |

## LICENSE

MIT


CHANGELOG.md
# [1.0.0](https://github.com/xg4/text2image/compare/v0.3.3...v1.0.0) (2022-06-30)

### Features

- add loadImage method ([14a8a52](https://github.com/xg4/text2image/commit/14a8a52c4cd5aebd027fb8c72a9d6587fb182411))
- export default instance ([913f442](https://github.com/xg4/text2image/commit/913f4426fac4fdb9c2f68e48b6e8aa531445d146))
- reset temporary options at the end of the call ([e3889ad](https://github.com/xg4/text2image/commit/e3889adf0d7e8be57f6978ff84904fd2fb81dc75))

### BREAKING CHANGES

- cannot save the option to createURL method and toDataURL method
- export class,export default instance
- remove static create, createMask methods


src/index.ts
interface Options {
  fontSize: number | string
  color: string
  fontFamily: string
  fontWeight: string | number
  type: string
  quality: number
  text: string
  gradient?: [number, string][]
  alpha: number
}

const defaultOptions: Options = {
  fontSize: '30px',
  color: '#000000',
  fontFamily: 'Arial',
  fontWeight: 'normal',
  type: 'image/png',
  quality: 0.92,
  text: '',
  alpha: 0.3,
}

export class Text2Image {
  get width() {
    return this.c.width
  }

  get height() {
    return this.c.height
  }

  get font() {
    return `${this.options.fontWeight} ${this.options.fontSize} ${this.options.fontFamily}`
  }

  private defaultOptions: Options

  private currentOptions: Partial<Options> | null

  private mask: HTMLImageElement | null

  private c: HTMLCanvasElement

  private ctx: CanvasRenderingContext2D

  constructor(options?: Partial<Options>) {
    this.c = document.createElement('canvas')
    const ctx = this.c.getContext('2d')
    if (!ctx) {
      throw new Error('Failed to create canvas context')
    }

    this.ctx = ctx

    this.mask = null
    this.currentOptions = null

    this.defaultOptions = {
      ...defaultOptions,
    }

    this.setDefaultOptions(options)
  }

  get options(): Options {
    return { ...this.defaultOptions, ...this.currentOptions }
  }

  setDefaultOptions(options?: Partial<Options>) {
    Object.assign(this.defaultOptions, options)
  }

  resetDefaultOptions() {
    this.defaultOptions = { ...defaultOptions }
  }

  setMask(image?: HTMLImageElement) {
    this.mask = image || null
  }

  async loadImage(url: string): Promise<HTMLImageElement> {
    return new Promise((resolve, reject) => {
      const image = new Image()
      image.crossOrigin = 'anonymous'
      image.onload = () => {
        this.setMask(image)
        resolve(image)
      }
      image.onerror = reject
      image.src = url
    })
  }

  toDataURL(text?: string | Partial<Options>) {
    if (text) {
      this.currentOptions = this.parseOptions(text)
    }

    this.draw()

    const result = this.c.toDataURL(this.options.type, this.options.quality)
    this.currentOptions = null

    return result
  }

  createURL(text?: string | Partial<Options>): Promise<string> {
    return new Promise((resolve) => {
      if (text) {
        this.currentOptions = this.parseOptions(text)
      }

      this.draw()

      this.c.toBlob(
        (blob) => {
          if (!blob) {
            throw new Error('Failed to create blob')
          }
          resolve(URL.createObjectURL(blob))
          this.currentOptions = null
        },
        this.options.type,
        this.options.quality
      )
    })
  }

  destroyURL(url: string) {
    return URL.revokeObjectURL(url)
  }

  parseOptions(text: string | Partial<Options>): Partial<Options> {
    return typeof text === 'string' ? { text } : text
  }

  private drawMask() {
    if (!this.mask) {
      return
    }

    this.ctx.save()
    const { width: w, height: h } = this
    const { width: mw, height: mh } = this.mask

    const scale = h / mh
    const computedHeight = Math.min(h, mh)
    const computedWidth = mw * scale

    this.ctx.globalAlpha = this.options.alpha
    this.ctx.drawImage(
      this.mask,
      w / 2 - computedWidth / 2,
      h / 2 - computedHeight / 2,
      computedWidth,
      computedHeight
    )
    this.ctx.restore()
  }

  private drawText() {
    this.ctx.save()

    if (this.options.gradient) {
      const gradient = this.ctx.createLinearGradient(0, 0, this.width, 0)
      this.options.gradient.forEach((v) => {
        gradient.addColorStop(...v)
      })
      this.ctx.fillStyle = gradient
    } else {
      this.ctx.fillStyle = this.options.color
    }

    this.ctx.font = this.font
    this.ctx.textAlign = 'center'
    this.ctx.textBaseline = 'middle'
    this.ctx.translate(this.width / 2, this.height / 2 + this.height * 0.08)
    this.ctx.fillText(this.options.text, 0, 0)
    this.ctx.restore()
  }

  private measureText() {
    this.ctx.save()
    this.ctx.font = this.font
    const width = this.ctx.measureText(this.options.text).width
    this.ctx.restore()
    return {
      width,
      height: +this.options.fontSize.toString().replace(/\D/g, ''),
    }
  }

  private draw() {
    if (!this.options.text) {
      throw new Error('Invalid text')
    }

    const { width, height } = this.measureText()
    this.c.height = height
    this.c.width = width
    this.ctx.clearRect(0, 0, this.width, this.height)

    if (!this.width || !this.height) {
      throw new TypeError(
        `Invalid width or height, width:${this.width} height:${this.height}`
      )
    }

    // draw
    this.drawText()

    this.drawMask()
  }
}

const text2image = new Text2Image()

export default text2image
